core/project_settings.cpp

```diff
@@ -86,6 +86,7 @@ String ProjectSettings::localize_path(const String &p_path) const {
 		// DirAccess::get_current_dir() is not guaranteed to return a path that with a trailing '/',
 		// so we must make sure we have it as well in order to compare with 'res_path'.
 		cwd = cwd.plus_file("");
+		cwd = cwd.substr(0, cwd.length() - 1);
```

Comment:
This defeats the documented purpose of the line just above...
0639946c72ba6632bc3b0953d64f644af328e5e6
CC @neikeq

Reply:
It does defeat its purpose. I'll look for examples of paths 0639946c72ba6632bc3b0953d64f644af328e5e6 fixes.
This looks weird though. `plus_file` makes sure not to duplicate paths. Is this windows?

Reply:
The main problem is that by it returning a trailing forward slash, it results in double forward slashes when combined with the rest of the file, and my initial fix which involved trimming the the forward slash in the other parts of the localize_path function resulted in certain path getting mangled. In this attempted fix, I kept the 'plus_file' method on the assumption that the commented seemed to imply that we could gurantee the string ending with a forward slash, resulting in it not having anything to cut off. It is definetly a bug though, as and attempt to load a script will end up with it being loaded with a double forward slash if it's not already been cached, and results in it not saving properly. Are there any ideas for a better way to fix it, since this bug seems like such a common problem that I'm surprised I haven't seen anyone else mention it.

Reply:
Sorry for the long text, but I think this is worth explaining in full detail for the record, otherwise even I will forget this. If you don't care about details, just jump to the solutions.

Let's suppose our project directory is `/path/to/project/`.

### Prior to 0639946c72ba6632bc3b0953d64f644af328e5e6

- `/path/to/project/File.scn` **->** `/path/to/project/File.scn` **WRONG**
- `/path/to/project/Foo/File.scn` **->**
  - If the `Foo` sub-directory exist: `res://Foo/File.scn` **RIGHT**
  - If the `Foo` sub-directory **doesn't** exist: `/path/to/project/Foo/File.scn` **WRONG**

### After 0639946c72ba6632bc3b0953d64f644af328e5e6

- `/path/to/project/File.scn` **->** `res:///File.scn` **ALMOST RIGHT**

Works but it introduces the extra `/` bug was introduced.

- `/path/to/project/Foo/File.scn` **->**
  - If the `Foo` sub-directory exist: `res://Foo//File.scn` **ALMOST RIGHT**
  - If the `Foo` sub-directory **doesn't** exist: `res:///Foo/File.scn` **ALMOST RIGHT**

Works but the same extra `/` bug but in different directories depending of whether  `Foo` exists or not.

### After #34279 (this PR)

Same results as _**"Prior to 0639946c72ba6632bc3b0953d64f644af328e5e6"**_. Basically it reverts the fix from 0639946c72ba6632bc3b0953d64f644af328e5e6, which also reverts the extra `/` bug it introduces.

### What went wrong with #33380

The now reverted PR #33380 introduced bug #33913:

- Both `foo/bar/spam.txt` and `/path/to/project/foo/bar/spam.txt` **->** `res://foobarspam.txt` **WRONG**

What the issue doesn't mention is that this bug only happens if both the `foo` and `bar` sub-directories don't exist.

In brief `localize_path` is doing the following:

```
(Localized path to first found dir) + (Non-existent sub-dirs, if any) + (File, if any)
```

Prior to #33380:

```C++
("res://") + ("/foo" + "/bar") + ("/spam.txt") == "res:///foo/bar/spam.txt"; // extra `/` bug
("res://foo/") + ("/bar") + ("/spam.txt") == "res://foo//bar/spam.txt"; // extra `/` bug
("res://foo/bar/") + ("/spam.txt") == "res://foo/bar//spam.txt"; // extra `/` bug
```

After #33380:

```C++
("res://") + ("foo" + "bar") + ("spam.txt") == "res://foobarspam.txt"; // 2 missing `/`
("res://foo/") + ("bar") + ("spam.txt") == "res://foo/barspam.txt"; // 1 missing `/`
("res://foo/bar/") + ("spam.txt") == "res://foo/bar/spam.txt"; // right
```

## Solutions

### Only strip the `/` from the file (or non existent sub-dir) if the parent directory ends with `/`

```diff
 @@ -107,6 +107,10 @@ String ProjectSettings::localize_path(const String &p_path) const {
 		if (plocal == "") {
 			return "";
 		};
+
+		// Only strip the starting '/' from 'path' if its parent ('plocal') ends with '/'
+		if (plocal[plocal.length() - 1] == '/')
+			sep += 1;
+
 		return plocal + path.substr(sep, path.size() - sep);
 	};
 }
```

Does the following:

``` C++
("res://") + ("foo" + "/bar") + ("/spam.txt") == "res://foo/bar/spam.txt"; // right
("res://foo/") + ("bar") + ("/spam.txt") == "res://foo/bar/spam.txt"; // right
("res://foo/bar/") + ("spam.txt") == "res://foo/bar/spam.txt"; // right
```

### Same but comply with simplify_path

If we want to return a simplified path, then we have to make `localize_path` result never end with a `/` (unless the result is exactly "res://").

```diff
 @@ -91,7 +91,9 @@ String ProjectSettings::localize_path(const String &p_path) const {
 			return p_path;
 		};
 
-		return cwd.replace_first(res_path, "res://");
+		// Strip forward slash at the end, except if result == "res://"
+		String result = cwd.replace_first(res_path, "res://");
+		return result.length() == 6 /* length of res:// */ ? result : result.substr(0, result.length() - 1);
 	} else {
 
 		memdelete(dir);
@@ -107,6 +109,11 @@ String ProjectSettings::localize_path(const String &p_path) const {
 		if (plocal == "") {
 			return "";
 		};
+
+		// Only strip the starting '/' from 'path' if its parent ('plocal') ends with '/'.
+		// This is needed because 'plocal' may be exactly "res://".
+		if (plocal[plocal.length() - 1] == '/')
+			sep += 1;
+
		return plocal + path.substr(sep, path.size() - sep);
	};
}
```

Does the following:

``` C++
("res://") + ("foo" + "/bar") + ("/spam.txt") == "res://foo/bar/spam.txt"; // right
("res://foo") + ("/bar") + ("/spam.txt") == "res://foo/bar/spam.txt"; // right
("res://foo/bar") + ("/spam.txt") == "res://foo/bar/spam.txt"; // right
```

Results (same for both solutions):

- `/path/to/project/File.scn` **->** `res://File.scn` **RIGHT**
- `/path/to/project/Foo/File.scn` **->**
  - If the `Foo` sub-directory exist: `res://Foo/File.scn` **RIGHT**
  - If the `Foo` sub-directory **doesn't** exist: `res://Foo/File.scn` **RIGHT**
- `foo/bar/spam.txt` **->** `res://foobarspam.txt` **RIGHT**

Difference between both solutions:

- Both `Bar` and `/path/to/project/Bar` **->**
  - If the `Bar` sub-directory exists:
    - Solution 1: `res://Bar/`
    - Solution 2: `res://Bar`
  - If the `Bar` **doesn't** sub-directory exists:
    - Both solutions 1 and 2: `res://Bar`

The second solution matches `simplify_path` and is consistent in the output whether the sub-directory exists or not.

Reply:
A third simpler and safer but potentially slower solution:

```diff
@@ -91,7 +91,7 @@ String ProjectSettings::localize_path(const String &p_path) const {
 			return p_path;
 		};
 
-		return cwd.replace_first(res_path, "res://");
+		return cwd.replace_first(res_path, "res://").simplify_path();
 	} else {
 
 		memdelete(dir);
@@ -107,7 +107,7 @@ String ProjectSettings::localize_path(const String &p_path) const {
 		if (plocal == "") {
 			return "";
 		};
-		return plocal + path.substr(sep, path.size() - sep);
+		return plocal.plus_file(path.substr(sep, path.size() - sep)).simplify_path();
 	};
 }
 
```

Results are the same as the second solution, as it complies with `simplify_path` (because it uses it...).

Reply:
@neikeq I can change to one of these solutions no problem, but I'd like to leave it up to you guys to decide if there's a preference for a particular fix you'd like to see. Both seem fairly viable to me